Workflows: try a backport changelog

ellatrix opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

This PR introduces a test that checks if an entry with a core backport PR has been added to the backport changelog.

Entries are added as files to prevent merge/rebase conflicts. The expectation would be one file per core backport PR, which can contain a list of multiple Gutenberg PRs.

The files can be sorted per release: 6.6 changes go in the 6.6 folder.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

The goal of this PR is to make sure that core backport PRs are created before a PR that requires a backport is merged into GB. This encourages PR authors to think about how it will be absorbed into core early, increases visibility, and allows us to merge the core PRs as soon as possible, which prevents a "queue" of many PRs to be backported (that may depend on one another).

The PR author will have to create a backport at some point anyway. I think it's more productive to do it sooner while it’s fresh and while you’re looking at the core files that you’re filtering/overriding.

If there's an existing (and open) core backport PR that a change is related to, it's ok to use that existing core PR. So if you wish to make multiple successive PHP changes, it's fine to start a backport PR and use that for all these related PRs. This makes it easier for others to know which PRs belong together (normally when a backport issue is created, it's really hard to figure out which PRs belong together and can go into the same backport PR).

Also note that you don't need to open a backport PR as soon as you create a GB PR, it should just be done before merging to trunk. Maybe we can skip running this test for draft PRs.

**Probably we'll go through some time where we have remaining backports (#61226) and it's not possible to create a backport PR without previous changes. We are now two weeks from Beta 1, so we should prioritise getting those remaining backports merged as soon as possible.**

If this works well, I can see it replace the backport issue entirely, or it becomes just of formality where we double check all the changes, but almost no work is needed.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

This script is inspired by the components changelog: we check a special changelog whenever a PHP change is made.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

* I'm first running this PR without any PHP changes, to make sure that the test does not run.
* Then I'll make a PHP change to make sure the test fails. (skip confirmed)
* I'll then add an entry to changelog until it passes. (works: https://github.com/WordPress/gutenberg/actions/runs/9154811094/job/25166026310)
* Finally I'll remove the PHP change and changelog entry again.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

ellatrix committed:
Workflows: try a backport changelog

ellatrix committed:
Make a PHP change

ellatrix committed:
Try adding an entry

ellatrix committed:
Don't require md links, they are auto-linked

vcanales commented:
> ...make sure that core backport PRs are created as soon as a change is merged in GB that requires a backport.

I understand this as the PR has to be **created and added to the changelog before the change is merged** into PR, no? Apologies if this sounds pedantic, I'm just a bit confused by the order of operations :) 

> Maybe this PR obsoletes the other PHP change detection test.

Perhaps a combination rather than a replacement? I think it makes sense to have only one comment addressing PHP changes that need to be Backported and added to the changelog.

ellatrix commented:
Changed the wording

> Perhaps a combination rather than a replacement? I think it makes sense to have only one comment addressing PHP changes that need to be Backported and added to the changelog.

Yes, we can keep both if it's useful

ellatrix committed:
Remove test change

ellatrix committed:
Use files to prevent conflicts

ellatrix committed:
Test making a change

ellatrix committed:
Test wrong file name

ellatrix committed:
Add correct entry

ellatrix committed:
Check without link

ellatrix committed:
File per core backport pr

ellatrix committed:
Test if it passes

ellatrix committed:
Fix example

ellatrix committed:
Remove test change

priethor commented on .github/workflows/check-backport-changelog.yml:
```diff
@@ -0,0 +1,58 @@
+
+                  # Confirm that there is an entry containing the Gutenberg PR link
+                  if [[ -z "${changelog_file}" ]]; then
+                    echo "Please create a core backport PR and add a file with the name <core-pr-number>.md in the $changelog_folder folder with the core backport PR URL and a list item with this PR URL."
+                    echo "If changes are related to an existing, open core PR, you may add this PR URL to the core PR's file."
```
Can we extract these repeated strings to a variable?

ellatrix replied:
I removed the first check, the second check does the same basically.

ellatrix committed:
Move example

priethor commented on .github/workflows/check-backport-changelog.yml:
```diff
@@ -0,0 +1,58 @@
+
+                  # Confirm that there is an entry containing the Gutenberg PR link
+                  if [[ -z "${changelog_file}" ]]; then
+                    echo "Please create a core backport PR and add a file with the name <core-pr-number>.md in the $changelog_folder folder with the core backport PR URL and a list item with this PR URL."
```
> with the name <core-pr-number>.md

Why not use `${changelog_file}` to say exactly what the file's name should be?

ellatrix replied:
We don't know the core PR number in advance

priethor commented on backport-changelog/readme.md:
```diff
@@ -0,0 +1,3 @@
+# Core Backport Changelog
+
+Any PR that makes changes to be backported to core should log a core PR here. It's possible to have multiple Gutenberg PRs link to a single core backport PR. The core backport PR can remain open as long as wanted/needed. The entries are sorted by core release, and each entry should be an md file with the Gutenberg PR number as the file name, and the link to the core backport PR as file content. F ...
```
```suggestion
Any PR that makes changes to be backported to core should log a core PR here. It's possible to have multiple Gutenberg PRs link to a single core backport PR. The core backport PR can remain open as long as wanted/needed. The entries are sorted by core release, and each entry should be an md file with the Gutenberg PR number as the file name, and the link to the core backport PR as file content (see `<core-pr-number>.md` as an example). Files are used to avoid rebase conflicts.
```

ellatrix replied:
Refactored this and added an inline example.

priethor reviewed (approved):
It's a good, low-risk first attempt, and I see no reason not to try it and iterate later.

ellatrix committed:
Update docs, remove unnecessary check

ellatrix committed:
Test again after changes

ellatrix committed:
Add more exceptions, test modifying both checked and ignored paths

ellatrix committed:
Test just with an ignored path

ellatrix committed:
Test again (should pass)

ellatrix committed:
Test again (should fail 2nd check)

ellatrix committed:
Remove test

ellatrix committed:
Add packages php files + exceptions

ramonjd commented:
> This PR introduces a test that checks if an entry with a core backport PR has been added to the backport changelog.

Hi there! 

Thanks for this - going to help with release season for sure.

Quick question: is there a way to opt-out for PHP changes that don't require backports, e.g., PRs that sync from Core -> Gutenberg

Or this one I'm working on

- https://github.com/WordPress/gutenberg/pull/61561

ramonjd commented:
> Quick question: is there a way to opt-out for PHP changes that don't require backports, e.g., PRs that sync from Core -> Gutenberg

A "try" branch here skipping via a label:

- https://github.com/WordPress/gutenberg/pull/61808
